base/compiler/tfuncs.jl

```diff
@@ -484,14 +484,27 @@ function typeof_tfunc(@nospecialize(t))
 end
 add_tfunc(typeof, 1, 1, typeof_tfunc, 0)
 
-function typeassert_tfunc(@nospecialize(v), @nospecialize(t))
-    t = instanceof_tfunc(t)[1]
-    t === Any && return v
+function typeassert_instance(@nospecialize(v), @nospecialize(t))
     if isa(v, Const)
         if !has_free_typevars(t) && !isa(v.val, t)
             return Bottom
         end
         return v
+    elseif isa(v, PartialStruct)
+        has_free_typevars(t) && return v
+        widev = widenconst(v)
+        newT = typeintersect(widev, t)
+        if newT === widev
```

Comment:
```suggestion
        if newT == widev
```

Reply:
I know these seem almost the same (since `.typ` is usually concrete), but the type-system knows that too, and can handle the exceptions better

Reply:
I figured the case where typeintersect would return something that's `==` to one of the inputs, but not `===` to one of the inputs would be pretty rare, and in the case where it doesn't, using `===` would bypass the useless sometype query that we don't really care about.